app/src/main/java/io/homeassistant/companion/android/sensors/LastAppSensorManager.kt

```diff
@@ -0,0 +1,88 @@
+package io.homeassistant.companion.android.sensors
+
+import android.Manifest
+import android.app.usage.UsageStats
+import android.app.usage.UsageStatsManager
+import android.content.Context
+import android.content.pm.PackageManager
+import android.os.Build
+import android.util.Log
+import androidx.annotation.RequiresApi
+import io.homeassistant.companion.android.common.sensors.SensorManager
+import java.util.TreeMap
+import io.homeassistant.companion.android.common.R as commonR
+
+class LastAppSensorManager : SensorManager {
+    companion object {
+        private const val TAG = "LastApp"
+
+        val last_used = SensorManager.BasicSensor(
+            "last_used_app",
+            "sensor",
+            commonR.string.basic_sensor_name_last_used_app,
+            commonR.string.sensor_description_last_used_app
+        )
+    }
+
+    override fun docsLink(): String {
+        return "https://companion.home-assistant.io/docs/core/sensors#last-used-app-sensor"
+    }
+    override val enabledByDefault: Boolean
+        get() = false
+    override val name: Int
+        get() = commonR.string.sensor_name_last_app
+
+    override fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
+        return listOf(last_used)
+    }
+
+    override fun hasSensor(context: Context): Boolean {
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
+    }
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun requiredPermissions(sensorId: String): Array<String> {
+        return arrayOf(Manifest.permission.PACKAGE_USAGE_STATS)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.LOLLIPOP_MR1)
+    override fun requestSensorUpdate(
+        context: Context
+    ) {
+        updateLastApp(context)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.LOLLIPOP_MR1)
+    private fun updateLastApp(context: Context) {
+        if (!isEnabled(context, last_used.id))
+            return
+
+        val usageStats = context.getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
+        val current = System.currentTimeMillis()
+        val appList = usageStats.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, current - 1000 * 1000, current)
+        val sortedMap = TreeMap<Long, UsageStats>()
+        if (!appList.isNullOrEmpty() && appList.size > 0) {
+            for (item in appList) {
+                sortedMap[item.lastTimeUsed] = item
+            }
+        }
+
+        var lastApp = sortedMap.maxByOrNull { it.key }?.value?.packageName ?: "none"

```

Comment:
This is what I was suggesting:
```suggestion
        val lastApp = usageStats.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, current - 1000 * 1000, current).maxByOrNull { it.lastTimeUsed }?.packageName ?: "none"
```

Reply:
thank you for that :) that is much better!